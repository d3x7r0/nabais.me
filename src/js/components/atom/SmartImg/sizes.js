const IMAGE_SIZES = [
  120,
  160,
  240,
  320,
  480,
  768,
  920,
  1280,
  1600,
  1920,
  2560,
  3840,
]

/**
 * @param {{ min?: number, max?: number}} options
 * @param {number[]} candidates
 * @returns {number[]}
 */
export function calculateSizes(options = {}, candidates = IMAGE_SIZES) {
  const { min, max } = options

  let result = Array.from(candidates)

  if (min) {
    result = result.filter(entry => min < entry)
  }

  if (max) {
    result = result.filter((entry, idx, arr) => {
      const prev = arr[idx - 1] || 0

      return max >= entry || max >= prev
    })
  }

  return result
}
