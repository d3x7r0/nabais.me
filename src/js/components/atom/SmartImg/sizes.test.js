import { describe, expect, it } from 'vitest'

import { calculateSizes } from './sizes'

describe('components/atom/SmartImg', () => {
  describe('calculateSizes', () => {
    it('handles min value correctly', () => {
      const result1 = calculateSizes({ min: 100 })
      expect(result1).toEqual([
        120,
        160,
        240,
        320,
        480,
        768,
        920,
        1280,
        1600,
        1920,
        2560,
        3840,
      ])

      const result2 = calculateSizes({ min: 1000 })
      expect(result2).toEqual([
        1280,
        1600,
        1920,
        2560,
        3840,
      ])

      const result3 = calculateSizes({ min: 4000 })
      expect(result3).toEqual([])
    })

    it('handles max value correctly', () => {
      const result1 = calculateSizes({ max: 100 })
      expect(result1).toEqual([
        120,
      ])

      const result2 = calculateSizes({ max: 1280 })
      expect(result2).toEqual([
        120,
        160,
        240,
        320,
        480,
        768,
        920,
        1280,
        1600,
      ])

      const result3 = calculateSizes({ max: 1800 })
      expect(result3).toEqual([
        120,
        160,
        240,
        320,
        480,
        768,
        920,
        1280,
        1600,
        1920,
      ])

      const result4 = calculateSizes({ max: 4000 })
      expect(result4).toEqual([
        120,
        160,
        240,
        320,
        480,
        768,
        920,
        1280,
        1600,
        1920,
        2560,
        3840,
      ])
    })

    it('handles min and max values correctly', () => {
      const result1 = calculateSizes({ min: 0, max: 100 })
      expect(result1).toEqual([
        120,
      ])

      const result2 = calculateSizes({ min: 300, max: 1300 })
      expect(result2).toEqual([
        320,
        480,
        768,
        920,
        1280,
        1600,
      ])

      const result3 = calculateSizes({ min: 800, max: 4000 })
      expect(result3).toEqual([
        920,
        1280,
        1600,
        1920,
        2560,
        3840,
      ])
    })
  })
})
