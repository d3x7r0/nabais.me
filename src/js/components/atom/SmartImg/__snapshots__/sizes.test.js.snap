// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/atom/SmartImg calculateSizes handles max value correctly 1`] = `
[
  120,
]
`;

exports[`components/atom/SmartImg calculateSizes handles max value correctly 2`] = `
[
  120,
  160,
  240,
  320,
  480,
  768,
  920,
  1280,
  1600,
]
`;

exports[`components/atom/SmartImg calculateSizes handles max value correctly 3`] = `
[
  120,
  160,
  240,
  320,
  480,
  768,
  920,
  1280,
  1600,
  1920,
]
`;

exports[`components/atom/SmartImg calculateSizes handles max value correctly 4`] = `
[
  120,
  160,
  240,
  320,
  480,
  768,
  920,
  1280,
  1600,
  1920,
  2560,
  3840,
]
`;

exports[`components/atom/SmartImg calculateSizes handles min and max values correctly 1`] = `
[
  120,
]
`;

exports[`components/atom/SmartImg calculateSizes handles min and max values correctly 2`] = `
[
  320,
  480,
  768,
  920,
  1280,
  1600,
]
`;

exports[`components/atom/SmartImg calculateSizes handles min and max values correctly 3`] = `
[
  920,
  1280,
  1600,
  1920,
  2560,
  3840,
]
`;

exports[`components/atom/SmartImg calculateSizes handles min value correctly 1`] = `
[
  120,
  160,
  240,
  320,
  480,
  768,
  920,
  1280,
  1600,
  1920,
  2560,
  3840,
]
`;

exports[`components/atom/SmartImg calculateSizes handles min value correctly 2`] = `
[
  1280,
  1600,
  1920,
  2560,
  3840,
]
`;

exports[`components/atom/SmartImg calculateSizes handles min value correctly 3`] = `[]`;
